examples/opensora_cai/opensora/models/layers/blocks.py

```diff
@@ -491,3 +621,122 @@ def construct(self, labels, train, force_drop_ids=None):
             labels = self.token_drop(labels, force_drop_ids)
         embeddings = self.embedding_table(labels)
         return embeddings
+
+
+class SizeEmbedder(nn.Cell):
+    def __init__(self, hidden_size, frequency_embedding_size=256):
+        super().__init__()
+        self.mlp = nn.SequentialCell(
+            nn.Dense(frequency_embedding_size, hidden_size),
+            nn.SiLU(),
+            nn.Dense(hidden_size, hidden_size),
+        )
+        self.frequency_embedding_size = frequency_embedding_size
+        self.outdim = hidden_size
+
+    def construct(self, s: Tensor, bs: Tensor) -> Tensor:
+        if s.ndim == 1:
+            s = s[:, None]
+        assert s.ndim == 2
+        if s.shape[0] != bs:
+            s = s.repeat(bs // s.shape[0], axis=0)
+            assert s.shape[0] == bs
+        b, dims = s.shape[0], s.shape[1]
+        s = s.reshape(b * dims)  # b d -> (b d)
+        s_freq = TimestepEmbedder.timestep_embedding(s, self.frequency_embedding_size)
+        s_emb = self.mlp(s_freq)
+        return s_emb.reshape(b, dims * self.outdim)  # (b d) d2 -> b (d d2)
+
+
+class CaptionEmbedder(nn.Cell):
+    """
+    Embeds class labels into vector representations. Also handles label dropout for classifier-free guidance.
+    """
+
+    # FIXME: rm nn.GELU instantiate for parallel training
+    def __init__(self, in_channels, hidden_size, uncond_prob, act_layer=nn.GELU, token_num=120):
+        super().__init__()
+
+        self.y_proj = Mlp(
+            in_features=in_channels, hidden_features=hidden_size, out_features=hidden_size, act_layer=act_layer, drop=0
+        )
+
+        y_embedding = ops.randn(token_num, in_channels) / in_channels**0.5
+        # just for token dropping replacement, not learnable
+        self.y_embedding = Parameter(Tensor(y_embedding, dtype=ms.float32), requires_grad=False)
+
+        self.uncond_prob = uncond_prob
+
+    def token_drop(self, caption, force_drop_ids=None):
+        """
+        Drops labels to enable classifier-free guidance.
+        """
+        if force_drop_ids is None:
+            drop_ids = ops.rand(caption.shape[0]) < self.uncond_prob
+        else:
+            drop_ids = force_drop_ids == 1
+
+        # manually expand dims to avoid infer-shape bug in ms2.3 daily
+        caption = ops.where(
+            drop_ids[:, None, None, None], self.y_embedding[None, None, :, :], caption.to(self.y_embedding.dtype)
+        )
+
+        return caption
+
+    def construct(self, caption, train, force_drop_ids=None):
+        if train:
+            assert caption.shape[2:] == self.y_embedding.shape
+        use_dropout = self.uncond_prob > 0
+        if (train and use_dropout) or (force_drop_ids is not None):
+            caption = self.token_drop(caption, force_drop_ids)
+        caption = self.y_proj(caption)
+        return caption
+
+
+class PositionEmbedding2D(nn.Cell):
```

Comment:
Is this module support dynamic `dim` during training?

Reply:
No, but I think the support can be added, if necessary.